lib/Sema/CSApply.cpp

```diff
@@ -2475,28 +2475,34 @@ namespace {
                                          baseTyUnwrapped,
                                          memberName,
                                          defaultMemberLookupOptions);
-          
-          // Lookup didn't find anything, so return
+
+          // Filter out any functions, instance members, enum cases with
+          // associated values or variables whose type does not match the
+          // contextual type.
+          results.filter([&](const LookupResultEntry entry, bool isOuter) {
+            if (auto member = entry.getValueDecl()) {
+              if (isa<FuncDecl>(member))
+                return false;
+              if (member->isInstanceMember())
+                return false;
+              if (auto EED = dyn_cast<EnumElementDecl>(member)) {
+                return !EED->hasAssociatedValues();
+              }
+              if (auto VD = dyn_cast<VarDecl>(member)) {
+                auto baseType = DSCE->getType()->lookThroughAllOptionalTypes();
+                return VD->getInterfaceType()->isEqual(baseType);
+              }
+            }
+
+            return true;
```

Comment:
Do we think there'll be any other results, or should this be `llvm_unreachable`?

Reply:
We might have a `typealias none = <whatever>` :-)

Reply:
True. If one of those is there, what would the right thing to do be?

Reply:
Hmm I thought we'd get a redeclaration error, but seems like we don't get one if you have a computed property. So, I think the right thing to do here is to simply filter them out, otherwise we're gonna trip the warning:

```swift
enum Foo {
  case bar
  var none: Foo { . bar }
  struct none {}
}

let foo: Foo? = .none // warning 
```

Reply:
That's an instance property rather than a static property. I was thinking of the case where you *just* have the type, though.

Reply:
If we just have `struct none {}` then we'd trip the warning as well. Anyway, since it doesn't happen now because it gets filtered out.

Reply:
I have added some more tests (it's getting really edge case-y now 😅).